kratos/sources/table_accessor.cpp

```diff
@@ -0,0 +1,87 @@
+//                   Carlos Roig
+//
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/table_accessor.h"
+#include "includes/properties.h"
+
+namespace Kratos
+{
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double TableAccessor::GetValue(
+    const Variable<double>& rVariable,
+    const Properties& rProperties,
+    const GeometryType& rGeometry,
+    const Vector& rShapeFunctionVector,
+    const ProcessInfo& rProcessInfo
+    ) const
+{
+    return GetValueFromTable(*mpInputVariable, rVariable, rProperties, rGeometry, rShapeFunctionVector, rProcessInfo);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double TableAccessor::GetValueFromTable(
+        const Variable<double> &rIndependentVariable,
+        const Variable<double> &rDependentVariable,
+        const Properties& rProperties,
+        const GeometryType& rGeometry,
+        const Vector& rShapeFunctionVector,
+        const ProcessInfo& rProcessInfo) const
+{
+    double independent_at_gauss = 0.0;
+    if (mInputVariableType == static_cast<int>(InputVariableType::NodalHistorical)) {
+        KRATOS_ERROR_IF_NOT(rGeometry[0].SolutionStepsDataHas(rIndependentVariable)) << "The Variable " << rIndependentVariable.Name() << " is not available at the nodes of the Geometry to retrieve Table values." << std::endl;
+        for (SizeType i = 0; i < rShapeFunctionVector.size(); ++i) {
+            KRATOS_DEBUG_ERROR_IF_NOT(rGeometry[i].SolutionStepsDataHas(rIndependentVariable)) << "The Variable " << rIndependentVariable.Name() << " is not available at the nodes of the Geometry to retrieve Table values." << std::endl;
+            const double nodal_value = rGeometry[i].FastGetSolutionStepValue(rIndependentVariable);
+            independent_at_gauss += nodal_value * rShapeFunctionVector[i];
+        }
+    } else if (mInputVariableType == static_cast<int>(InputVariableType::NodalNonHistorical)) {
+        KRATOS_ERROR_IF_NOT(rGeometry[0].Has(rIndependentVariable)) << "The Variable " << rIndependentVariable.Name() << " is not available at the nodes of the Geometry to retrieve Table values." << std::endl;
+        for (SizeType i = 0; i < rShapeFunctionVector.size(); ++i) {
+            KRATOS_DEBUG_ERROR_IF_NOT(rGeometry[i].Has(rIndependentVariable)) << "The Variable " << rIndependentVariable.Name() << " is not available at the nodes of the Geometry to retrieve Table values." << std::endl;
+            const double nodal_value = rGeometry[i].GetValue(rIndependentVariable);
+            independent_at_gauss += nodal_value * rShapeFunctionVector[i];
+        }
+    } else if (mInputVariableType == static_cast<int>(InputVariableType::ElementalNonHistorical)) {
```

Comment:
as they commented, if you store the enum, here you will also not need to cast it